ui/src/App.js

```diff
@@ -176,31 +209,44 @@ class App extends Component {
       let parts = action.removedValue.value.split("=");
       this.updateFacets(parts[0], parts[1], false);
     } else if (action.action == "select-option") {
+      let option = action.option;
       // Drop-down row was clicked.
       let newExtraFacetEsFieldNames = this.state.extraFacetEsFieldNames;
-      newExtraFacetEsFieldNames.push(action.option.esFieldName);
-      this.setState({ extraFacetEsFieldNames: newExtraFacetEsFieldNames });
+      newExtraFacetEsFieldNames.push(option.esFieldName);
+
+      let currentFilterMap = this.state.selectedFacetValues;
+      if (this.state.enableSearchValues && option.facetValue != "") {
+        currentFilterMap = this.updateSelectedFacetValues(
+          option.esFieldName,
+          option.facetValue,
+          true
+        );
+      }
+
       this.facetsApi.facetsGet(
         {
-          filter: this.filterMapToArray(this.state.selectedFacetValues),
+          filter: this.filterMapToArray(currentFilterMap),
           extraFacets: newExtraFacetEsFieldNames
         },
-        this.facetsCallback
+        function(error, data) {
+          this.facetsCallback(error, data);
+          this.setState({
+            selectedFacetValues: currentFilterMap,
+            extraFacetEsFieldNames: newExtraFacetEsFieldNames
+          });
+        }.bind(this)
       );
     }
   }
 
-  /**
-   * Updates the selection for a single facet value and refreshes the facets data from the server.
-   * */
-  updateFacets(esFieldName, facetValue, isSelected) {
+  updateSelectedFacetValues(esFieldName, facetValue, isSelected) {
     let currentFilterMap = this.state.selectedFacetValues;
```

Comment:
currentFilterMap -> allFacetValues
currentFacetValues -> facetValuesForField

As is, 259 is confusing -- why return currentFilterMap rather than newFilterMap